src/libcore/slice/sort.rs

```diff
@@ -701,3 +701,65 @@ pub fn quicksort<T, F>(v: &mut [T], mut is_less: F)
 
     recurse(v, &mut is_less, None, limit);
 }
+
+fn recurse_at_index<'a, T, F>(v: &'a mut [T], index: usize, is_less: &mut F, pred: Option<&'a T>)
+    where F: FnMut(&T, &T) -> bool
+{
+    // For slices of up to this length it's probably faster to simply sort them.
+    const MAX_INSERTION: usize = 10;
+    if v.len() <= MAX_INSERTION {
+        insertion_sort(v, is_less);
+        return;
+    }
+
+    // Choose a pivot
+    let (pivot, _) = choose_pivot(v, is_less);
+
+    // If the chosen pivot is equal to the predecessor, then it's the smallest element in the
+    // slice. Partition the slice into elements equal to and elements greater than the pivot.
+    // This case is usually hit when the slice contains many duplicate elements.
+    if let Some(p) = pred {
+        if !is_less(p, &v[pivot]) {
+            let mid = partition_equal(v, pivot, is_less);
+
+            // If we've passed our index, then we're good.
+            if mid > index {
+                return;
+            }
+
+            // Otherwise, continue sorting elements greater than the pivot.
+            recurse_at_index(&mut v[mid..], index - mid, is_less, None);
+            return;
+        }
+    }
+
+    let (mid, _) = partition(v, pivot, is_less);
+
+    // Split the slice into `left`, `pivot`, and `right`.
+    let (left, right) = {v}.split_at_mut(mid);
+    let (pivot, right) = right.split_at_mut(1);
+    let pivot = &pivot[0];
+
+    // If mid == index, then we're done, since partition() guaranteed that all elements after mid
+    // are greater than or equal to mid.
+    if mid < index {
+        recurse_at_index(right, index - mid - 1, is_less, Some(pivot));
+    } else if mid > index {
+        recurse_at_index(left, index, is_less, pred);
```

Comment:
Any reason why we have recursion here? Looks like this function should be easy to turn into a loop.

Reply:
I changed it to a loop. PTAL.

Reply:
Looks good, thanks! 👍